API Change: DMA - callback improvement

tbursztyka opened the pull request:
- **Problem Description:**

DMA callback do not provide the DMA device's callback caller as parameter, forcing users to use the callback_arg parameter instead for this purpose, which in turn limits its usage (they cannot provide some private data, or have to mangle thing together storing the DMA dev pointer into the private data).

 - **Proposed Change (detailed):**

All other API provide the callback's caller device pointer, and thus adding it as first parameter to the DMA callback.

 - **Dependencies:**

All users of function need to adapt their callbacks accordingly.

Provide the DMA device instance to the callback.

Commit:
api: dma: Pass the dma device pointer as parameter to callback

Normalize callback like other APIs.
This clarifies the callback usage. It will also help to switch all
device instance to constants.

Fixes #26923

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers: dma: Apply DMA callback change to relevant drivers

Now the dma device instance is passed as parameter to the callback.

Fixes #26923

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
tests: Apply DMA callback change to relevant tests

Now the dma device instance is passed as parameter to the callback.

Fixes #26923

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

tbursztyka commented:
@carlescufi let's bring this last one also to next API meeting

Commit:
doc: Add a DMA API change entry to release-notes for 2.4 version

Documenting the change.

Fixes #26923

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

pabigot reviewed (approved):
Looks good.

I was a little concerned about whether `void *callback_arg` should be moved to the last argument and renamed `user_data`, but based on the scanty documentation for this I don't think its role is equivalent to user data so there's no motivation to move it.

carlescufi commented:
@nordic-krch and @anangl  could you please take a look?

henrikbrixandersen reviewed (approved):
LGTM

Commit:
dma: Generalize dma callback signature by adding a dedicated typedef

Until now, each and every driver had to redeclare the type to store such
pointer, now they will be using this typedef.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers: dma: Align all drivers using dma_callback_t type

No need to re-declare such signature anymore.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
dma: Callback's user data has to be called user_data

s/callback_arg/user_data

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers: dma: Align all drivers using user_data for callback

s/callback_arg/user_data

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
tests: Align all drivers using user_data for callback

s/callback_arg/user_data

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>
